grammars/tree-sitter-php.cson

```diff
@@ -0,0 +1,322 @@
+  'string': [
+    {match: '^"', scopes: 'string.quoted.double.php'},
+    'string.quoted.single.php'
+  ]
+  # FIXME this isn't working, how to scope the ' and " in strings?
+  # 'string > "\'"': 'punctuation.definition.string.begin.php'
+  # NOTE the following is in language-ruby
+  # "string > '\"':nth-child(0)": 'punctuation.definition.string.begin'
+  # "string > '\"':nth-child(1)": 'punctuation.definition.string.end'
+  'heredoc': [
+    {match: "^[\r\n\s]*<<<'", scopes: 'string.unquoted.nowdoc.php'},
+    'string.unquoted.heredoc.php'
+  ]
+  'comment': [
+      {match: '^/\\*\\*', scopes: 'comment.block.documentation.phpdoc.php'},
+      {match: '^//', scopes: 'comment.line.double-slash.php'},
+      'comment.block.php'
+  ]
+  # FIXME these are not working
+  # 'comment > "//"': 'punctuation.definition.comment.php'
+  # 'comment > "/*"': 'punctuation.definition.comment.php'
+  # 'comment > "/**"': 'punctuation.definition.comment.php'
+  # 'comment > "*/"': 'punctuation.definition.comment.php'
+
+
+  # copied from TM grammar, removed T/F/null b/c those are named nodes in TS
+  'qualified_name > name': {match: /^(__(FILE|DIR|FUNCTION|CLASS|METHOD|LINE|NAMESPACE)__|ON|OFF|YES|NO|NL|BR|TAB)$/i, scopes: 'constant.language.php'}
+
+  'echo_statement > parenthesized_expression > "("': 'punctuation.definition.arguments.begin.bracket.round.php'
+  'echo_statement > parenthesized_expression > ")"': 'punctuation.definition.arguments.end.bracket.round.php'
+  'parenthesized_expression > "("': 'punctuation.definition.begin.bracket.round.php'
+  'parenthesized_expression > ")"': 'punctuation.definition.end.bracket.round.php'
+  'object_creation_expression > arguments > "("': 'punctuation.definition.begin.bracket.round.php'
+  'object_creation_expression > arguments > ")"': 'punctuation.definition.end.bracket.round.php'
+  'arguments > "("': 'punctuation.definition.arguments.begin.bracket.round.php'
+  'arguments > ")"': 'punctuation.definition.arguments.end.bracket.round.php'
+  'formal_parameters > "("': 'punctuation.definition.parameters.begin.bracket.round.php'
+  'formal_parameters > ")"': 'punctuation.definition.parameters.end.bracket.round.php'
+  'anonymous_function_use_clause > "("': 'punctuation.definition.parameters.begin.bracket.round.php'
+  'anonymous_function_use_clause > ")"': 'punctuation.definition.parameters.end.bracket.round.php'
+  'catch_clause > "("': 'punctuation.definition.parameters.begin.bracket.round.php'
+  'catch_clause > ")"': 'punctuation.definition.parameters.end.bracket.round.php'
+  'array_creation_expression > "("': 'punctuation.definition.array.begin.bracket.round.php'
+  'array_creation_expression > ")"': 'punctuation.definition.array.end.bracket.round.php'
+  'object_creation_expression > declaration_list > "{"': 'punctuation.definition.class.begin.bracket.curly.php'
+  'object_creation_expression > declaration_list > "}"': 'punctuation.definition.class.end.bracket.curly.php'
```

Comment:
This matches the TM grammar, but feels like it's unnecessarily specific. For example, the [JS grammar](https://github.com/atom/language-javascript/blob/master/grammars/tree-sitter-javascript.cson#L164-L170) has all parens as just `punctuation.definition.parameters.(begin|end).bracket.round`. I would like to change all of these to just `punctuation.definition.(begin|end).bracket.round.php` and see how that affects themes.

Reply:
Most of uses lists all of them: https://grep.app/search?q=punctuation%5C.definition%5C.%5Ba-z.%5D%2B%5C.bracket%5C.%5Ba-z.%5D%2B%5C.php&regexp=true

If is working, I'd personally leave this as is, but I don't think that change will break anything. Up to you.

Reply:
We could also later make PR to replace this in both TS and TM, and gather feedback.

Reply:
Ah, OK, makes sense. For now I've rearranged them so that all of the array/arguments/parameters are least grouped together. Thanks!